pvanalytics/features/shading.py

```diff
@@ -0,0 +1,338 @@
+        A pandas Series with index `index` and values from the image.
+    """
+    return pd.Series(image.flatten(), index=image_index)
+
+
+def _smooth(image):
+    """Smooth `image` by local mean filtering.
+
+    Local means are computed in a 3x3 square surrounding each pixel.
+    """
+    return ndimage.uniform_filter(image, size=15)
+
+
+def _prepare_images(ghi, clearsky, daytime, interval):
+    """Prepare data as images.
+
+    Performs pre-processing steps on `ghi` and `clearsky` before
+    returning images for use in the shadow detection algorithm.
+
+    Returns
+    -------
+    ghi_image : np.ndarray
+        Image form of `ghi`
+    clearsky_image : np.ndarray
+        Image form of `clearsky`
+    clouds_image : np.ndarray
+        Image of the cloudy periods in `ghi`
+    image_times : pandas.DatetimeIndex
+        Index for the data included in the returned images. Leading
+        and trailing days with incomplete data are not included in the
+        image, these times are needed to build a Series from the image
+        later on.
+
+    """
+    # Fill missing days by interpolation. Interpolation is limited to
+    # times with valid data before and after; missing data at the
+    # beginning or end of the series is not filled in.
+    image_width = 1440 // interval
+    ghi = ghi.interpolate(limit_area='inside')
+    # drop incomplete days.
+    ghi = ghi[ghi.resample('D').transform('count') == image_width]
+    image_times = ghi.index
+    ghi_image = _to_image(ghi.to_numpy(), image_width)
+    scaled_ghi = (ghi * 1000) / _smooth(ghi_image).max()
+    scaled_clearsky = (clearsky * 1000) / clearsky.max()
+    scaled_clearsky = scaled_clearsky.reindex_like(scaled_ghi)
+    daytime = daytime.reindex_like(scaled_ghi)
+    # Detect clouds.
+    clouds = _detect_clouds(scaled_ghi, scaled_clearsky, '50T')
+    # Interpolate across days (i.e. along columns) to remove clouds
+    # replace clouds with nans
+    #
+    # This could probably be done directly with scipy.interpolate.inter1d,
+    # but the easiest approach is to turn the image into a dataframe and
+    # interpolate along the columns.
+    cloudless_image = ghi_image.copy()
+    cloudless_image[_to_image(clouds.to_numpy(), image_width)] = np.nan
+    clouds_image = ghi_image.copy()
+    clouds_image[~_to_image(clouds.to_numpy(), image_width)] = np.nan
+    ghi_image = pd.DataFrame(cloudless_image).interpolate(
+        axis=0,
+        limit_direction='both'
+    ).to_numpy()
+    # set night to nan
+    ghi_image[~_to_image(daytime.to_numpy(), image_width)] = np.nan
+    return (
+        ghi_image,
+        _to_image(scaled_clearsky.to_numpy(), image_width),
+        clouds_image,
+        image_times
+    )
+
+
+def _detect_clouds(ghi, clearsky_ghi, window='50T'):
+    """Use :py:func:`pvanalytics.clearsky.reno` to detect clouds"""
+    clouds = pvlib.clearsky.detect_clearsky(
+        ghi,
+        clearsky_ghi,
+        ghi.index,
+        window_length=10
```

Comment:
`window='50T'` and `window_length=10`? Wondering if `window_length=50` gets what we want directly and avoids `clouds.rolling(window).sum()`

Reply:
This function is probably affected by [changes](https://github.com/pvlib/pvlib-python/pull/1074) made in pvlib v0.8., where the intervals are changed from left-labeled to centered.

Reply:
The way I read the original code is that it used a window of 10 to do the cloud detection, then did a convolution with a 50 minute window to keep only 50-minute or longer cloudy periods. I tried `window_length=50`, but it substantially changes the output.

For pvlib >= 0.8.1 we need `align='left'` I guess. Might be time to bump the minimum version of pvlib to 0.8.1, but that should probably be done separately from this PR since it may require other changes as well.

Reply:
> The way I read the original code is that it used a window of 10 to do the cloud detection, then did a convolution with a 50 minute window to keep only 50-minute or longer cloudy periods.

That sounds right.

>  I tried `window_length=50`, but it substantially changes the output.

Agree. I withdraw my suggestion. Maybe change the parameter to `duration='50T'` to avoid confusion with the `window_length` parameter for the pvlib function.
 
> For pvlib >= 0.8.1 we need `align='left'` I guess. Might be time to bump the minimum version of pvlib to 0.8.1, but that should probably be done separately from this PR since it may require other changes as well.

My comment was made when thinking about matching the Matlab code output. Maybe that's not necessary and this implementation can live with centered intervals for determining clear-sky times.